virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1251 @@
+		printf("Missing pod ID\n");
+		goto err;
+	}
+
+	if (!params->runtime_path) {
+		printf("Missing runtime path\n");
+		goto err;
+	}
+
+	return 0;
+
+err:
+	free_watcher_params(params);
+	return -1;
+}
+
+/* 
+ * Enter the network namespace.
+ *
+ * Entering the provided network namespace since the point of this binary is
+ * to monitor any network change happening inside a specific network namespace.
+ */
+int enter_netns(const char *netns_path)
+{
+	int fd;
+
+	fd = open(netns_path, O_RDONLY);
+	if (fd == -1) {
+		printf("Failed opening network ns %s: %s\n",
+				netns_path, strerror(errno));
+		return -1;
+	}
+
+	if (setns(fd, 0) == -1) {
+		printf("Failed to join network ns %s: %s\n",
+				netns_path, strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * Open the netlink socket with appropriate parameters, and return the file
+ * descriptor. It will listen for any message related to a network interface
+ * change (add/delete/update), an IP address (IPv4) change (add/delete/update)
+ * or a route (IPv4) change (add/delete/update).
+ */
+int open_netlink()
+{
+	int sock;
+	struct sockaddr_nl sa;
+
+	memset(&sa, 0, sizeof(sa));
+
+	sa.nl_family = AF_NETLINK;
+	sa.nl_pid = getpid();
+	sa.nl_groups = RTMGRP_LINK |
+		RTMGRP_IPV4_IFADDR |
+		RTMGRP_IPV4_ROUTE;
+
+	sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (sock == -1) {
+		printf("Failed creating netlink socket: %s\n",
+				strerror(errno));
+		return -1;
+	}
+
+	if (bind(sock, (struct sockaddr *) &sa, sizeof(sa)) == -1) {
+		printf("Failed binding netlink socket: %s\n",
+				strerror(errno));
+		return -1;
+	}
+
+	return sock;
+}
+
+/*
+ * Execute the runtime call from a child process.
+ */
+int fork_runtime_call(char *params[])
+{
+	int ret;
+	int wstatus;
+	int exit_code;
+
+	int pid = fork();
+	if (pid < 0) {
+		printf("Could not spawn the child: %s\n", strerror(errno));
+		return -1;
+	} else if (!pid) {
+		execvp(params[0], params);
+
+		/* This part of the code should not be reached */
+		printf("Failed execvp() runtime\n");
+		return -1;
+	}
+
+	if (waitpid(pid, &wstatus, 0) == -1) {
+		printf("Failed waitpid(): %s\n", strerror(errno));
+		return -1;
+	}
+
+	if (WIFEXITED(wstatus)) {
+		exit_code = WEXITSTATUS(wstatus);
+		printf("Runtime exit code %d\n", exit_code);
+		if (!exit_code) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * Convert a binary IP address.
+ * 
+ * It's the caller responsibility to free the memory.
+ */
+char* ip_addr_to_string(unsigned char family, void *attr)
+{
+	int buf_len;
+	char *buf;
+
+	if (family == AF_INET6) {
+		buf_len = (INET6_ADDRSTRLEN + 1) * sizeof(char);
+	} else {
+		buf_len = (INET_ADDRSTRLEN + 1) * sizeof(char);
+	}
+
+	buf = (char *) malloc(buf_len);
+
+	return (char *) inet_ntop(family, attr, buf, buf_len);
+}
+
+/*
+ * Get network interface name from index.
+ * 
+ * It's the caller responsibility to free the memory.
+ */
+char* iface_idx_to_name(int idx)
+{
+	int buf_len = 32 * sizeof(char);
```

Comment:
What is `32`?

Reply:
Nevermind, I have just read the doc https://linux.die.net/man/3/if_indextoname and it clearly says that we should allocate a buffer of __at least__ `IF_NAMESIZE` (being 16). I think we can stick to that for now.